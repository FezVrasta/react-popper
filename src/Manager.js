// @flow
import React, { Component, type Node } from 'react';
import createContext, { type Context } from 'create-react-context';

export const ManagerContext: Context<{
  getReferenceRef?: (?HTMLElement) => void,
  referenceNode?: ?HTMLElement,
}> = createContext({});

type ManagerProps = {
  children: Node,
};
type ManagerState = {
  referenceNode?: ?HTMLElement,
};

export default class Manager extends Component<ManagerProps, ManagerState> {
  state = {};

  getReferenceRef = (referenceNode: ?HTMLElement) =>
    this.setState({ referenceNode });

  render() {
    return (
      <ManagerContext.Provider
        value={{
          referenceNode: this.state.referenceNode,
          getReferenceRef: this.getReferenceRef,
        }}
      >
        {this.props.children}
      </ManagerContext.Provider>
    );
  }
}
