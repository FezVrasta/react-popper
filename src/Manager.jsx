import { Component, createElement } from 'react'
import PropTypes from 'prop-types'

class Manager extends Component {
  static childContextTypes = process.env.NODE_ENV !== 'production'
    ? {
        popperManager: PropTypes.object.isRequired,
      }
    : {
        popperManager: () => {},
      }

  static propTypes = {
    tag: PropTypes.oneOfType([PropTypes.string, PropTypes.bool]),
    children: PropTypes.oneOfType([PropTypes.node, PropTypes.func]),
  }

  static defaultProps = {
    tag: 'div',
  }

  getChildContext() {
    return {
      popperManager: {
        setTargetNode: this._setTargetNode,
        getTargetNode: this._getTargetNode,
      },
    }
  }

  _setTargetNode = node => {
    this._targetNode = node
  }

  _getTargetNode = () => {
    return this._targetNode
  }

  render() {
    const { tag, children, ...restProps } = this.props
    if (tag !== false) {
      return createElement(tag, restProps, children)
    } else {
      return children
    }
  }
}

export default Manager
