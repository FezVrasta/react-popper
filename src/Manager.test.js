// @flow
import React from 'react';
import { mount } from 'enzyme';

// Public API
import { Manager } from '.';

// Private API
import { ManagerContext } from './Manager';

describe('Manager component', () => {
  it('renders the expected markup', () => {
    const wrapper = mount(
      <Manager>
        <div id="target" />
        <div id="popper" />
      </Manager>
    );
    expect(wrapper).toMatchSnapshot();
  });

  it('provides the related context', () => {
    const Target = () => null;
    const referenceNode = document.createElement('div');

    const wrapper = mount(
      <Manager>
        <ManagerContext.Consumer>
          {({ getReferenceRef, referenceNode }) => (
            <Target
              getReferenceRef={getReferenceRef}
              referenceNode={referenceNode}
            />
          )}
        </ManagerContext.Consumer>
      </Manager>
    );

    wrapper.find(Target).prop('getReferenceRef')(referenceNode);
    wrapper.update();
    expect(wrapper.find(Target).prop('referenceNode')).toBe(referenceNode);
  });
});

describe('TargetNodeContext', () => {
  it('provides proper default values', () => {
    const Target = () => null;
    const wrapper = mount(
      <div>
        <ManagerContext.Consumer>
          {({ getReferenceRef, referenceNode }) => (
            <Target
              getReferenceRef={getReferenceRef}
              referenceNode={referenceNode}
            />
          )}
        </ManagerContext.Consumer>
      </div>
    );

    expect(wrapper.find(Target).prop('getReferenceRef')).toBeUndefined();
  });
});
