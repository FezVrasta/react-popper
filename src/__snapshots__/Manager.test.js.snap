// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Manager component renders the expected markup 1`] = `
<Manager>
  <div
    id="target"
  />
  <div
    id="popper"
  />
</Manager>
`;
